### ISSUE TITLE
Better --api XML output

### ISSUE THREAD

**Issue by AngeloG** [languagetool-org/languagetool#251]
I noticed that the XML output of calls to the command line tool with `--api` output some weirdly formatted XML. First off, every paragraph outputs as its own document, resulting in an output like this: (manually indented for readability)

```XML
<?xml version="1.0" encoding="UTF-8"?>
<matches software="LanguageTool" version="2.9" buildDate="2015-03-30 09:24">
  <language shortname="en" name="English"/>
</matches>
<?xml version="1.0" encoding="UTF-8"?>
<matches software="LanguageTool" version="2.9" buildDate="2015-03-30 09:24">
  <language shortname="en" name="English"/>
</matches>
```

I'm guessing it's done to satisfy the compatibility with some sort of application this can be plugged in? Is there any reason why it's not done like the following:

```XML
<?xml version="1.0" encoding="UTF-8"?>
<paragraphs>
  <matches software="LanguageTool" version="2.9" buildDate="2015-03-30 09:24">
    <language shortname="en" name="English"/>
  </matches>
  <matches software="LanguageTool" version="2.9" buildDate="2015-03-30 09:24">
    <language shortname="en" name="English"/>
  </matches>
</paragraphs>
```

It also appears as if there's an extra `</matches>` at the end of the XML output:

```XML
<?xml version="1.0" encoding="UTF-8"?>
<matches software="LanguageTool" version="2.9" buildDate="2015-03-30 09:24">
<language shortname="en" name="English"/>
</matches>
</matches>
<!--
Time: 2325ms for 34 sentences (14.6 sentences/sec)
-->
```

**Comment by danielnaber** [comment]
This is a bug, it happens when the file is >64,000 characters. As a workaround, use smaller files for now. I'll see how we can fix this.

**Comment by AngeloG** [comment]
The file I'm inputting (via stdin) is definitely not that large. You are referring to the extra closing tag, correct?

**Comment by arysin** [comment]
Note the part of the code that's reading the whole file in the memory has a
problem of running cross-sentence rules across paragraphs.
The line-by-line code does not have that problem.

Andriy

2015-04-13 13:16 GMT-04:00 Angelo Geels <notifications@github.com>:

**Comment by danielnaber** [comment]
@AngeloG Could you send us the exact parameters you call? I could only reproduce the problem with large files (the problem with "every paragraph gets its own XML doc").

**Comment by AngeloG** [comment]
@danielnaber I'm calling it with this: `--api -m en -adl -` And then write the paragraphs to stdin. (I'm not sure whether it's with `\r\n` or `\n`, or if that actually matters, but it might be worth checking?)

**Comment by danielnaber** [comment]
@AngeloG Sorry, the problem is indeed not (only) the file size but the use of `-` for input. I'm going to analyze the bug further.

**Comment by danielnaber** [comment]
@arysin Do you have an example? I would have thought it was the other way round.

**Comment by milekpl** [comment]
If you have the `-` for input, the large file mode is activated automatically.

**Comment by danielnaber** [comment]
I'm still not sure: why don't we simply load the whole file and then check it, without all the logic around loading? Okay, that way nobody will be able to check a 16GB file on their laptop, but do we really need to support such cases?

**Comment by danielnaber** [comment]
@AngeloG This should be fixed now. If you rely on the builds, it should be in the next nightly snapshot at https://languagetool.org/download/snapshots/?C=M;O=D

**Comment by AngeloG** [comment]
Awesome! Nice to see. Will check it out soon!

**Comment by milekpl** [comment]
@danielnaber: I did some tests on big corpora just to see the number of matches and to build some statistics. I needed to process 4 GB files… Note also that one could want to use LT just as a tagger for really big files. So, basically, line-by-line mode (which is basically the core of the large file mode) might be retained for such cases. Maybe if we have a "out ouf memory exception", when trying to load the file, we should switch to the line mode automatically? This should be fairly easy to change.
